common/rpc/direct_peer_chooser.go

```diff
@@ -22,71 +22,118 @@ package rpc
 
 import (
 	"context"
+	"fmt"
 	"sync"
+	"sync/atomic"
 
 	"go.uber.org/yarpc/api/peer"
 	"go.uber.org/yarpc/api/transport"
 	"go.uber.org/yarpc/peer/direct"
+	"go.uber.org/yarpc/peer/hostport"
 	"go.uber.org/yarpc/yarpcerrors"
 
+	"github.com/uber/cadence/common"
 	"github.com/uber/cadence/common/dynamicconfig"
 	"github.com/uber/cadence/common/log"
 	"github.com/uber/cadence/common/log/tag"
 	"github.com/uber/cadence/common/membership"
+	"github.com/uber/cadence/common/metrics"
+)
+
+var (
+	noOpSubscriberInstance = &noOpSubscriber{}
 )
 
 // directPeerChooser is a peer.Chooser that chooses a peer based on the shard key.
 // Peers are managed by the peerList and peers are reused across multiple requests.
 type directPeerChooser struct {
+	status               int32
 	serviceName          string
 	logger               log.Logger
+	scope                metrics.Scope
 	t                    peer.Transport
 	enableConnRetainMode dynamicconfig.BoolPropertyFn
 	legacyChooser        peer.Chooser
 	legacyChooserErr     error
 	mu                   sync.RWMutex
+	peers                map[string]peer.Peer
 }
 
-func newDirectChooser(serviceName string, t peer.Transport, logger log.Logger, enableConnRetainMode dynamicconfig.BoolPropertyFn) *directPeerChooser {
+func newDirectChooser(
+	serviceName string,
+	t peer.Transport,
+	logger log.Logger,
+	metricsCl metrics.Client,
+	enableConnRetainMode dynamicconfig.BoolPropertyFn,
+) *directPeerChooser {
 	return &directPeerChooser{
 		serviceName:          serviceName,
-		logger:               logger,
+		logger:               logger.WithTags(tag.DestService(serviceName)),
+		scope:                metricsCl.Scope(metrics.P2PRPCPeerChooserScope).Tagged(metrics.DestServiceTag(serviceName)),
 		t:                    t,
 		enableConnRetainMode: enableConnRetainMode,
+		peers:                make(map[string]peer.Peer),
 	}
 }
 
 // Start statisfies the peer.Chooser interface.
-func (g *directPeerChooser) Start() error {
-	c, ok := g.getLegacyChooser()
-	if ok {
-		return c.Start()
+func (g *directPeerChooser) Start() (err error) {
+	if !atomic.CompareAndSwapInt32(&g.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return nil
 	}
 
-	return nil // no-op
+	defer func() { g.logger.Info("direct peer chooser started", tag.Error(err)) }()
+
+	if g.enableConnRetainMode != nil && !g.enableConnRetainMode() {
```

Comment:
if enableConnRetainMode is nil, we will default to connection retain mode.

Reply:
it's not supposed to be nil but since the check is there let me fix it so it defaults to legacy mode when nil